src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -117,13 +125,29 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
   def findCampaignsByUpdate(update: UpdateId): Future[Seq[Campaign]] =
     db.run(campaignRepo.findByUpdateAction(update))
 
+  /**
+   * Calculate campaign-wide statistic counters, also taking retry campaings
+   * into account if any exist.
+   */
   def campaignStats(campaignId: CampaignId): Future[CampaignStats] = for {
-    status    <- findClientCampaign(campaignId).map(_.status)
-    finished  <- countFinished(campaignId)
-    cancelled <- countCancelled(campaignId)
-    failed    <- failedDevices(campaignId)
-    stats     <- campaignStatsFor(campaignId)
-  } yield CampaignStats(campaignId, status, finished, failed, cancelled, stats)
+    mainCampaign <- campaignRepo.find(campaignId)
```

Comment:
There is a possibility the result will be inconsistent, as every query is running in its own transaction. I would suggest to refactor the code to build a DB action running in a transaction.